macrosynergy/pnl/historic_portfolio_volatility.py

```diff
@@ -120,31 +69,132 @@ def _estimate_variance_covariance(
     """
     # If weights is None, then the weights are equal
     if weights_arr is None:
+        # TODO weights_arr is never None in the current implementation
         weights_arr = np.ones(piv_ret.shape[0]) / piv_ret.shape[0]
 
     assert len(weights_arr) == len(piv_ret), "weights and window must have same length"
     assert not (
         piv_ret.isna().any().any()
     ), "NaN should have been removed by this stage"
+    # TODO why is N/A removed by this stage? Has it been replaced by zeros?
 
     cov_matr = np.zeros((len(piv_ret.columns), len(piv_ret.columns)))
 
-    for iB, cB in enumerate(piv_ret.columns):
-        for iA, cA in enumerate(piv_ret.columns[: iB + 1]):
-            est_vol = weighted_covariance(
-                x=piv_ret[cA].values,
-                y=piv_ret[cB].values,
+    for i_b, c_b in enumerate(piv_ret.columns):
+        for i_a, c_a in enumerate(piv_ret.columns[: i_b + 1]):
+            est_vol = _weighted_covariance(
+                x=piv_ret[c_a].values,
+                y=piv_ret[c_b].values,
                 w=weights_arr,
                 remove_zeros=remove_zeros,
             )
-            cov_matr[iA, iB] = est_vol  # TODO (iA, iB) is same as (iB, iA) 
-            cov_matr[iB, iA] = est_vol  # TODO (iA, iB) is same as (iB, iA)
+            cov_matr[i_a, i_b] = est_vol
+            if i_a != i_b:
+                cov_matr[i_b, i_a] = est_vol
 
     assert np.all((cov_matr.T == cov_matr) ^ np.isnan(cov_matr))
 
     return pd.DataFrame(cov_matr, index=piv_ret.columns, columns=piv_ret.columns)
 
 
+def _nan_ratio(x) -> float:
+    return (sum(np.isnan(x)) + sum(x == 0)) / len(x)
+
+
+def _mask_nan_series(
+    pivot_df: pd.DataFrame,
+    trigger_indices: pd.DatetimeIndex,
+    lback_periods: int,
+    nan_tolerance: float,
+) -> pd.DataFrame:
+    """
+    Drops series from the dataframe if the ratio of NaNs to non-NaNs exceeds the
+    nan_tolerance. Exists to allow greater flexibility in nan-processing.
+
+    :param <pd.DataFrame> pivot_df: the dataframe of the pivot table.
+    :param <pd.DatetimeIndex> trigger_indices: the DateTimeIndex of the trigger dates.
+    :param <int> lback_periods: the number of periods to use for filling NaNs.
+    :param <float> nan_tolerance: the maximum ratio of NaNs to non-NaNs in a lookback
+        window, if exceeded the resulting volatility is set to NaN. Default is 0.25.
+    """
+
+    for trigger_date in trigger_indices:
+        dtidx = pivot_df.index[pivot_df.index <= trigger_date][-lback_periods:]
+        col_mask = (pivot_df.loc[dtidx].apply(_nan_ratio) > nan_tolerance).values
+        pivot_df.loc[dtidx, col_mask] = np.nan
```

Comment:
Unless I am mistaken this won't work. We need to impose the max nan tolerance per trigger date.

As an example:
```python
batches: List[Tuple[datetime.date, pd.DataFrame]] = []
for trigger_date in trigger_indices:
    df = pivot_df.loc[pivot_df.index <= trigger_date, :].iloc[-lback_periods:]
    col_mask = (df.apply(_nan_ratio, axis=0) > nan_tolerance).values
    df.loc[:, col_mask] = np.nan
    batches.append((trigger_date, df))
```
Obviously this then requires us to modify the logic later down. Alternatively we need to impose the N/A logic at a later stage.